kernel/cooking.ml

```diff
@@ -256,7 +251,7 @@ let expand_subst0 cache expand_info abstr_ctx abstr_ausubst c =
   c
 
 let expand_subst cache c =
-  expand_subst0 cache.cache cache.info.expand_info cache.info.abstr_info.abstr_ctx cache.info.abstr_info.abstr_ausubst c
+  expand_subst0 cache.cache cache.info.expand_info (Lazy.force cache.named_ctx) cache.info.abstr_info.abstr_ausubst c
```

Comment:
`expand_subst0` has type 
~~~
(Univ.Instance.t * int list * int) RefTable.t ->
abstr_inst_info Names.Cmap.t * abstr_inst_info Names.Mindmap.t ->
('a, 'b) NamedDecl.pt list ->
Univ.Instance.t ->
Constr.constr ->
Constr.constr
~~~
so it doesn't care about the terms in the named_ctx.
In fact the named_ctx is only used through https://github.com/coq/coq/blob/fdc8569cf9a24bcd90e3db218f56999b78557808/kernel/cooking.ml#L140-L150 which just does `NamedDecl.get_id` so it could be replaced by an id list.

Reply:
I don't want to store more information than needed. It could go into the cache but there is little point. I've just reused the non-substituted instance from the info rather than forcing the substitution in the last version. It shouldn't matter though, almost all code path already force the context beforehand.

Reply:
I don't understand, how is named_context less info than id list?

Reply:
Sharing.

Reply:
what sharing?

Reply:
We can't store it into the info because we'd reallocate a list for every section constant / inductive. We could store it into the cache but then I don't see the gain because we're duplicating data and the new variant is not more efficient.

Reply:
I think I was looking at a previous version of the PR, now I'm confused.